book/src/dev/rfcs/0005-state-updates.md

```diff
@@ -520,22 +520,32 @@ to the in memory state, then we finalize all lowest height blocks that are
 past the reorg limit, finally we process any queued blocks and prune any that
 are now past the reorg limit.
 
-1. If the block itself exists in the finalized chain, it has already been successfully verified:
-  - broadcast `Ok(block.hash())` via `block.rsp_tx`, and return
+### `pub(super) fn commit_non_finalized_block(&mut self, block: QueuedBlock) -> block::Hash`
+
+1. If the block itself exists in a non-finalized chain, or the finalized chain,
+   it has already been successfully verified:
+     - broadcast `Ok(block.hash())` via `block.rsp_tx`, and return
 
-2. Run contextual validation on `block` against the finalized and non
+2. If the block itself exists in the queue:
+     - Create an extra receiver using `block.rsp_tx.subscribe`, and include that receiver
+       in the response future
+     - Drop the duplicate block
+
+3. Run contextual validation on `block` against the finalized and non
    finalized state
+     - contextual validation will reject blocks that are past the reorg limit,
```

Comment:
Did we want to start including specific contextual validation checks in this RFC? I was under the impression that this would end up in its own RFC.

Reply:
This is a note about satisfying the duplicate block handling invariant.
(And it's also a check included in the current implementation.)

That's why I included it here.